### PR TITLE
Fix ChangeState logic for limiting number of docs to commit in bulk

### LINKED ISSUE #11771: Workflows with no errors
**Impact of the bug**
Makes error handling for P&R very hard, as we rely on WM for error reports.

**Describe the bug**
Some worklfows (e.g. https://dmytro.web.cern.ch/dmytro/cmsprodmon/workflows.php?prep_id=task_BPH-RunIISummer20UL18GEN-00237) do not seem to have any errors. We do see some errors in the job log and WorkloadSummary (https://cmsweb.cern.ch/couchdb/workloadsummary/_design/WorkloadSummary/_show/histogramByWorkflow/pdmvserv_task_BPH-RunIISummer20UL18GEN-00237__v1_T_230926_104347_8763)

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes #11771 

#### Status
ready

#### Description
Previous code was buggy as we never incremented `countDocs`. In addition, we should actually have used `self.maxBulkCommit`, given that this is how the database connection instance is created and how we define the limit for commits in bulk.

In addition, now we also wrap `jsumdatabase` commits in a try/except, as we recently learned that those documents could be larger than the configured limit as well. Unfortunately, we have to deal with those slightly different, mainly for 2 reasons: 1) the document contains set() object, which is not serializable and thus we cannot get the full size of the document; 2) I picked 3 fields that can potentially be large (even though only `errors` has been a real example so far).

UPDATE: a list of 500 dicts for the `inputfiles` field adds around 100kB, so I think it's a good commitment.

It fixes a bug that I created in https://github.com/dmwm/WMCore/pull/11502

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
None

#### External dependencies / deployment changes
None

### CODE REVIEW

**Comment by amaltaro** [comment]
Things get better with the first commit, but now instead of failing to inject documents in `fwjrdatabase`, they fail at `jsumdatabase`. I am almost sure that the `lumis` attribute of the jobSummary document is the culprit. Investigating...

**Comment by amaltaro** [comment]
I was hopeful that after applying this patch (it's been applied to vocms0255), workflow job information would become available once again, but it did not happen for: `pdmvserv_task_PPS-Run3Summer22EEMiniAODv4-00177__v1_T_231101_141258_2773`

**Comment by amaltaro** [comment]
@todor-ivanov I am almost sure that the reason we do not see job related documents in CouchDB is because we are constantly failing to inject some documents into the local database, as their size is beyond the configured limits.

My first commit is actually fixing a logic that was put in place a couple of months ago, when injecting fwjr documents around this line:
https://github.com/dmwm/WMCore/blob/1702bc9a908d15c3f69bde1efe2e3a673ed8d6e4/src/python/WMCore/JobStateMachine/ChangeState.py#L394

it turns out that jsum documents can also fail, around this line:
https://github.com/dmwm/WMCore/blob/1702bc9a908d15c3f69bde1efe2e3a673ed8d6e4/src/python/WMCore/JobStateMachine/ChangeState.py#L518

where I saw the `errors` key having many megabytes. So I added a few checks to the python dictionary document before even queueing them.

As a short term fix, I would just deal with these document too large and truncate/reset some large fields.

Long term, we need to start fetching the outcome of these bulk operations and retrying only the specific document, as documented here: https://docs.couchdb.org/en/stable/api/database/bulk-api.html#api-db-bulk-docs-validation
This IMO is a larger development though and can potentially have side effects in places that we do not expect.

I just wanted to hear your thoughts on this and a review of this PR. One alternative for this PR could be: instead of checking the size of each `jobsummary` document before hand, to only check it in case we spot too large documents (same as it is implemented for fwjr, at the moment). Thanks

**Comment by amaltaro** [comment]
And we already have a ticket for the refactoring of CouchDB bulk operations:
https://github.com/dmwm/WMCore/issues/11576

File: src/python/WMCore/JobStateMachine/ChangeState.py [hunk 1]
```diff
@@ -82,6 +82,33 @@ def shrinkLargeFJR(couchDbInstance, sizeLimit):
...
+            fieldBytes = sys.getsizeof(json.dumps(doc[innerAttr]))
+            if fieldBytes > sizeLimit:
+                msg = f"Job summary id: {doc['_id']}, WMBS id: {doc['wmbsid']} "
+                msg += f"has {fieldBytes} bytes under the attr: {innerAttr}. Resetting it.."
+                logging.warning(msg)
+                if isinstance(doc[innerAttr], str):
+                    # keep only the first 5k characters
+                    doc[innerAttr] = doc[innerAttr][:5000]
+                    doc[innerAttr] += "... error message truncated ..."
+                elif isinstance(doc[innerAttr], list):
```
Comment (todor-ivanov) [inline src/python/WMCore/JobStateMachine/ChangeState.py:104]: Why not (similarly to the str field type) preserve the last, maybe hundreds, of elements from the list?

File: src/python/WMCore/JobStateMachine/ChangeState.py [hunk 2]
```diff
@@ -82,6 +82,33 @@ def shrinkLargeFJR(couchDbInstance, sizeLimit):
...
+                    doc[innerAttr] = []
+                elif isinstance(doc[innerAttr], dict):
```
Comment (todor-ivanov) [inline src/python/WMCore/JobStateMachine/ChangeState.py:106]: same comment here

File: src/python/WMCore/JobStateMachine/ChangeState.py [hunk 3]
```diff
@@ -387,11 +413,9 @@ def recordInCouch(self, jobs, newstate, oldstate, updatesummary=False):
                                 "archivestatus": archStatus,
                                 "fwjr": jsonFWJR,
                                 "type": "fwjr"}
-                if countDocs >= self.fwjrdatabase.getQueueSize():
+                if self.fwjrdatabase.getQueueSize() >= self.maxBulkCommit:
```
Comment (todor-ivanov) [inline src/python/WMCore/JobStateMachine/ChangeState.py:416]: Same question here

File: src/python/WMCore/JobStateMachine/ChangeState.py [hunk 4]
```diff
@@ -499,8 +523,19 @@ def recordInCouch(self, jobs, newstate, oldstate, updatesummary=False):
...
+
+                    if self.jsumdatabase.getQueueSize() >= self.maxBulkCommit:
+                        try:
+                            # TODO FIXME: https://github.com/dmwm/WMCore/issues/11576
+                            self.jsumdatabase.commit()
+                        except CouchRequestTooLargeError as exc:
+                            msg = "Failed to commit bulk of job summary documents to CouchDB."
+                            msg += f" Details: {str(exc)}"
+                            logging.warning(msg)
+                            shrinkLargeSummary(self.jsumdatabase, self.fwjrLimitSize)
```
Comment (todor-ivanov) [inline src/python/WMCore/JobStateMachine/ChangeState.py:535]: Ok, Let me see if I read it correctly: If we are in this situation, calling the `shringLargeSummary` function here is going to trigger a full scan of the local database and shrink only, the documents which are exceeding the configured size by cutting the fields that pass over some threshold? is My understanding correct?

**Review by todor-ivanov — COMMENTED** [review]
@amaltaro in general, it looks goo to me, but I've left 2 or 3 comments inline with some clarification questions. Could you  please take a look at them, before merging

File: src/python/WMCore/JobStateMachine/ChangeState.py (hunk 3 above)
Comment (amaltaro) [reply src/python/WMCore/JobStateMachine/ChangeState.py:416]: Same answer as above.

File: src/python/WMCore/JobStateMachine/ChangeState.py (hunk 1 above)
Comment (amaltaro) [reply src/python/WMCore/JobStateMachine/ChangeState.py:104]: I am going to answer it in the PR body itself to make it more readable.

**Comment by amaltaro** [comment]
@todor-ivanov answering your question on the `shrinkLargeSummary` function and why don't we truncate data for list/dict type as well. The short answer is, because it's more complicated and it increases the changes of having a final document that is still above the size limits.

The longer answer is, both `inputfiles` and `lumis` fields expect a `list` value data type (the dict is just a safeguard and we could potentially just remove that), where `inputfiles` has the following format:
```
 'inputfiles': [{'input_type': 'primaryFiles',
                 'lfn': '/store/unmerged/DMWM_Test/RelValProdMinBias/AODSIM/RECOPROD1_TaskChain_ProdMinBias_Nov2023_Val_Alanv2-v11/1930000/7A082F24-A37D-EE11-B99C-3CECEF0DDCE2.root'},
                {'input_type': 'primaryFiles',
                 'lfn': '/store/unmerged/DMWM_Test/RelValProdMinBias/AODSIM/RECOPROD1_TaskChain_ProdMinBias_Nov2023_Val_Alanv2-v11/1930000/6837B7E8-A27D-EE11-84A9-3CECEF5D9F50.root'},
```

and `lumis` looks like:
```
 'lumis': [{'1': {1: None,
                  ...
                  21: None}},
           {'1': {22: None,
                   ....
```

and I think for `inputfiles` we could indeed truncate it at 100 input files (or maybe 500 files, to cover merge jobs). But for the `lumis` field, it becomes tricky because we have dicts of dicts and their sizes can vary.

In addition, in the string data type case, it's simple to add a message saying that it has been truncated. While in the list/dict, I see no robust way to do that without affecting the data structure.

Another challenge with this PR is that the jobsummary document contains a/some set objects, which makes it unhashable and not possible to perform a json.dumps on the whole document. I am trying to figure which field is that though and see if anything better - not costly - can be done.

Please let me know if you have any suggestions.

**Review by todor-ivanov — APPROVED** [review]
Thank you @amaltaro, for those clarifications. All looks good to me.

**Comment by amaltaro** [comment]
@todor-ivanov thank you for your review. On top of it, I made the 6th commit (for bookkeepping: https://github.com/dmwm/WMCore/pull/11786/commits/ce3a313f9d8aafc61f811ecfa96f33736a05355d) which makes a few improvements to the function truncating/resetting. 

I was pretty sure I have had problems json dumping the whole document, but I haven't seen those again in my tests from yesterday and today, so I am giving this another try now.

**Comment by amaltaro** [comment]
As just noticed, I need to enclose L562:
```
        self.jsumdatabase.commit()
```
in a try/except block as well.

**Comment by amaltaro** [comment]
I have this patch updated in vocms0255 and now everything is looking good in the logs. This patch become a little more complex because the `errors` field isn't really a plain string, but a dict of lists - with potentially dictionary as items. 

I see no way to make the code simpler, but I am happy to hear of any suggestions.

@todor-ivanov if you review it again, commits 6 through 8 are supposed to be new code that you have not looked yet.

File: src/python/WMCore/JobStateMachine/ChangeState.py [hunk 5]
```diff
@@ -102,9 +112,12 @@ def shrinkLargeSummary(couchDbInstance, sizeLimit):
                 msg += f"has {fieldBytes} bytes under the attr: {innerAttr}. Truncating it.."
                 logging.warning(msg)
                 if innerAttr == "errors":
-                    # keep only the first 5k characters
-                    doc[innerAttr] = doc[innerAttr][:5000]
-                    doc[innerAttr] += "... error message truncated ..."
+                    # keep only the first 5k characters of the error detail
```
Comment (todor-ivanov) [inline src/python/WMCore/JobStateMachine/ChangeState.py:115]: At the moment the initial error dictionary has been created, was it assured, the next two fields used in the following two `for` cycles are actually iterrable (even though with zero length)?

**Review by todor-ivanov — APPROVED** [review]
Thanks @amaltaro I left only one comment inline, which is just a precaution.

File: src/python/WMCore/JobStateMachine/ChangeState.py (hunk 5 above)
Comment (amaltaro) [reply src/python/WMCore/JobStateMachine/ChangeState.py:115]: Yes, data type defaults to list.

**Pull request merged**